### PR TITLE
Align for kmhV3

### PR DESCRIPTION
**Pull request by tastynoob**
More accurate scheduling simulation and memory access behavior

### CODE REVIEW

**Comment by shinezyy** [comment]
It seems that 
7a8b4222e51a9804a20571bd2bd39ce91a5c0b52
and 
144acdf27a9fb3add7ef04b75369961c5e096535
can be merged into one commit?

**Comment by tastynoob** [comment]
> It seems that [7a8b422](https://github.com/OpenXiangShan/GEM5/commit/7a8b4222e51a9804a20571bd2bd39ce91a5c0b52) and [144acdf](https://github.com/OpenXiangShan/GEM5/commit/144acdf27a9fb3add7ef04b75369961c5e096535) can be merged into one commit?

fixed

File: src/arch/riscv/RiscvCPU.py [hunk 1]
```diff
@@ -79,13 +75,9 @@ class XiangshanECore(XiangshanCore):
     numPhysRMiscRegs = 40
 
 class XiangshanECore2Read(XiangshanCore):
-    fuPool = XSECore2ReadFUPool()
```
Comment (shinezyy) [inline src/arch/riscv/RiscvCPU.py:82]: Ecore shouldn't use the same fu config as kmh

File: src/cpu/o3/FUPool.py [hunk 2]
```diff
@@ -98,81 +55,156 @@ class SpecWakeupChannel(SimObject):
...
+#             IssuePort(fu=[SIMD_Unit()]),
+#             IssuePort(fu=[SIMD_Unit()]),
+#             IssuePort(fu=[SIMD_Unit()]),
+#             IssuePort(fu=[SIMD_Unit()]),
+#             IssuePort(fu=[SIMD_Unit()])
+#         ], scheduleToExecDelay=3),
+#     ]
+#     intSlotNum = 12
+#     fpSlotNum = 12
+#     xbarWakeup = True
```
Comment (shinezyy) [inline src/cpu/o3/FUPool.py]: plz remove useless code

File: src/cpu/o3/inst_queue.cc [hunk 3]
```diff
@@ -525,16 +511,118 @@ InstructionQueue::processFUCompletion(const DynInstPtr &inst, int fu_idx)
...
+            if (rs2 == 1) {
+                op_latency = 6;
+            } else if (rs1 == rs2) {
+                op_latency = 8;
+            } else if (lzc(std::labs(rs2)) - lzc(std::labs(rs1)) < 0) {
+                op_latency = 6;
+            } else {
+                op_latency = 8 + delay_ / 4;
+            }
+            return true;
```
Comment (shinezyy) [inline src/cpu/o3/inst_queue.cc:554]: plz add more comments for division latency computation and sqrt latency computation

File: src/cpu/o3/issue_queue.cc [hunk 4]
```diff
@@ -77,15 +99,67 @@ IssueQue::IssueQueStats::IssueQueStats(statistics::Group* parent, IssueQue* que,
...
+    }
+
+    if (same_fu) {
+        // we only allocate one ReadyQue
+        auto t = new ReadyQue;
+        readyQs.resize(outports, t);
+        auto& port = params.oports[0];
+        fuDescs.insert(fuDescs.begin(), port->fu.begin(), port->fu.end());
+    }
+    else {
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: } else {

File: src/cpu/o3/issue_queue.cc [hunk 5]
```diff
@@ -228,71 +309,90 @@ IssueQue::addIfReady(const DynInstPtr& inst)
...
 
 void
 IssueQue::selectInst()
 {
-    selectedInst.clear();
-    while (selectedInst.size() < inoutPorts && !readyInsts.empty()) {
-        auto& inst = readyInsts.top();
-        if (inst->canceled()) {
+    selectQ.clear();
+    for (int pi=0;pi<outports;pi++) {
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: plz format this line of code

File: src/cpu/o3/issue_queue.cc [hunk 6]
```diff
@@ -613,38 +749,28 @@ void
...
-            }
-        }
-    } else {
-        for (auto iq = iqs.rbegin(); iq != iqs.rend(); iq++) {
-            if ((*iq)->ready()) {
-                (*iq)->insert(inst);
-                inserted = true;
-                break;
-            }
+    std::random_shuffle(iqs.begin(), iqs.end());
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc:747]: add "TODO" here to let us remember to calibrate the insertion policy with RTL

File: src/cpu/o3/issue_queue.cc [hunk 7]
```diff
@@ -696,14 +831,21 @@ Scheduler::getInstToFU()
...
     uint32_t priority = getArbPriority(inst);
     uint32_t needed = inst->numSrcRegs();
-    slotOccupied += needed;
-    intSlot.push(Slot(priority, needed, inst));
+
+    if (inst->isFloating()) {
+        fpSlotOccupied += needed;
+        fpSlot.push(Slot(priority, needed, inst));
+    }
+    else if (inst->isInteger()) {
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: } else {

File: src/cpu/o3/issue_queue.hh [hunk 8]
```diff
@@ -51,23 +52,34 @@ class MemDepUnit;
...
     const int scheduleToExecDelay;
     const std::string iqname;
-    const std::vector<FUDesc *> fuDescs;
-
+    std::vector<FUDesc *> fuDescs;
+    std::vector<bool> opPipelined;
     int IQID = -1;
 
-    struct compare_priority
+    struct select_ploy
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.hh]: what is ploy?

File: src/cpu/o3/lsq_unit.cc [hunk 9]
```diff
@@ -70,6 +70,166 @@ namespace gem5
...
+{
+    assert(offset + size <= validMask.size());
+    for (uint64_t i = 0; i < size; ++i) {
+        blockDatas[offset + i] = datas[i];
+        validMask[offset + i] = true;
+    }
+}
+
+bool
+StoreBufferEntry::coverage(PacketPtr pkt, LSQ::LSQRequest *req)
```
Comment (shinezyy) [inline src/cpu/o3/lsq_unit.cc]: what does coverage mean here?

File: src/cpu/o3/lsq_unit.cc [hunk 10]
```diff
@@ -311,6 +422,9 @@ LSQUnit::LSQUnit(uint32_t lqEntries, uint32_t sqEntries, uint32_t sbufferEntries
       storeInFlight(false),
       stats(nullptr)
 {
+    sqFullUpperLimit = sqEntries - 4;
+    sqFullLowerLimit = sqFullUpperLimit - 4;
```
Comment (shinezyy) [inline src/cpu/o3/lsq_unit.cc:427]: what does "4" means here? define a constant for it ?

File: src/cpu/o3/lsq_unit.cc [hunk 11]
```diff
@@ -1232,15 +1373,34 @@ LSQUnit::storeBufferEvictToCache()
     if (storeBuffer.unsentSize() == 0) {
         return;
     }
-    if (storeBuffer.unsentSize() > sbufferEvictThreshold ||
-        storeBufferWritebackInactive > storeBufferInactiveThreshold || storeBufferFlushing) {
+
+    if (storeQueue.size() > sqFullUpperLimit) {
+        sqwillfull = true;
```
Comment (shinezyy) [inline src/cpu/o3/lsq_unit.cc]: var should be either snake case or camel case

File: src/cpu/o3/lsq_unit.hh [hunk 12]
```diff
@@ -304,33 +259,14 @@ class LSQUnit
...
-
-        bool coverage(PacketPtr pkt, LSQ::LSQRequest* req);
-    };
-
+    const int maxSQoffload = 2;
+    int sqFullUpperLimit = 0;
+    int sqFullLowerLimit = 0;
     bool storeBufferFlushing = false;
-
+    bool sqwillfull = false;
```
Comment (shinezyy) [inline src/cpu/o3/lsq_unit.hh]: member var should be camel case

**Review by shinezyy — CHANGES_REQUESTED** [review]

File: src/cpu/o3/issue_queue.cc [hunk 13]
```diff
@@ -77,15 +99,67 @@ IssueQue::IssueQueStats::IssueQueStats(statistics::Group* parent, IssueQue* que,
...
+            if (params.oports[i]->mask != params.oports[j]->mask) {
+                same_fu = false;
+            }
+            if (!same_fu && (params.oports[i]->mask & params.oports[j]->mask).any()) {
+                panic("%s: Found the conflict opClass in different FU, portid: %d and %d\n", iqname, i, j);
+            }
+        }
+    }
+    if (same_fu) {
+        warn("%s: Use one selector by multiple identical fus\n", iqname);
```
Comment (jueshiwenli) [inline src/cpu/o3/issue_queue.cc]: how about put the warn in the next same_fu?

File: src/cpu/o3/issue_queue.cc (hunk 4 above)
Comment (jueshiwenli) [inline src/cpu/o3/issue_queue.cc]: } else {

File: src/cpu/o3/issue_queue.cc [hunk 14]
```diff
@@ -696,14 +831,21 @@ Scheduler::getInstToFU()
...
     }
     uint32_t priority = getArbPriority(inst);
     uint32_t needed = inst->numSrcRegs();
-    slotOccupied += needed;
-    intSlot.push(Slot(priority, needed, inst));
+
+    if (inst->isFloating()) {
+        fpSlotOccupied += needed;
+        fpSlot.push(Slot(priority, needed, inst));
+    }
```
Comment (jueshiwenli) [inline src/cpu/o3/issue_queue.cc]: } elseif () {

**Pull request closed without merging**